docs/source/scheduled-queries.rst

```diff
@@ -0,0 +1,297 @@
+
+
+How do scheduled queries work?
+``````````````````````````````
+This system leverages two main components: AWS Lambda and AWS Step Functions.
+
+First, a CloudWatch scheduled event triggers the execution of a new AWS Step Function State Machine.
+This State Machine manages the lifecycle of Athena queries through the Lambda function. Its role
+responsibility is to execute the Lambda, wait a predefined window of time, and execute the Lambda again,
+repeating until the Lambda reports it is finshed.
+
+The Lambda function is a simple function that Starts Athena queries, caches their execution ids, checks
+on their execution status, and uploads results to StreamAlert via Kinesis. Instead of doing all of these
+steps in a blocking fashion and sleeping while it waits for Athena, it runs through all queries in a single
+nonblocking pass, and returns the result of its execution to the State Machine. Once all queries have
+completed and their results sent to StreamAlert, the Lambda returns a "done" flag to the State Machine,
+signalling that this job has been finished.
+
+
+
+Configuration
+-------------
+Scheduled Queries is configured via a single file, ``conf/scheduled_queries.json``.
+
+.. code-block::
+
+  {
+    "enabled": true,
+    "config": {
+      "destination_kinesis": "ryxias20200212_prod_streamalert",
+      "sfn_timeout_secs": 3600,
+      "sfn_wait_secs": 30
+    },
+    "packs": {
+      "hourly": {
+        "description": "Runs all hourly queries",
+        "schedule_expression": "rate(1 hour)"
+      },
+      "two_hour": {
+        "description": "Runs all queries run every two hours",
+        "schedule_expression": "rate(2 hours)"
+      },
+      "daily": {
+        "description": "Runs all daily queries",
+        "schedule_expression": "rate(24 hours)"
+      },
+      "two_day": {
+        "description": "Runs all queries that are run once every 2 days",
+        "schedule_expression": "rate(2 days)"
+      }
+    },
+    "lambda_config": {}
+  }
+
+* ``enabled`` — (bool) Pass `true` to activate ScheduledQueries. Leave `false` to disable.
+* ``config.destination_kinesis`` — (str) The name of the Kinesis stream to upload results to.
+* ``config.sfn_timeout_secs`` - (int) Max number of seconds for the state machine to execute.
+* ``config.sfn_wait_secs`` - (int) Time to wait between checks of query status.
+
+
+
+Query Packs
+```````````
+Query Packs are batches of scheduled Athena queries that are executed together.
+
+.. code-block::
+
+  "hourly": {
+    "description": "Runs all hourly queries",
+    "schedule_expression": "rate(1 hour)"
+  }
+
+* ``packs`` - (dict) The Keys of this dict are the **names** of the query packs. This is important later.
+  Each value is another dict with two keys:
+
+  - ``description`` - (str) A string summary of what queries belong in this pack.
+  - ``schedule_expression`` - (str) A CloudWatch schedule expression defining how frequently to execute this query pack.
```

Comment:
Does this mean that you have to define the schedule expression twice? Once here and again in the params section? Would it be possible to have it automagically pass that parameter based on the value set by schedule expression?

Reply:
No, you define it once. This is referencing the same document—I'll rewrite the documentation to be clearer

Reply:
Ah no, on second thought, I see what you mean... yeah that would be a big improvement in user experience

I think it would be possible but it would be a lot of complex work, because you have to handle really weird cases like `cron(38 */12 1,2,5 * *)` (omg). Additionally, some queries I've written run overlapping partitions. (Example: the BOX app queries ran 24 hours and scanned past 48 hours)

I think it might be possible to deal with this by storing a cursor in SSM or DynamoDB. Every execution, it understands which partitions have been scanned... and then continues on forward with unscanned partitions.

Ugh, but there's also some subtlety to the actual partition. Example; if you run a query at 3:05 PM, are you intending on scanning the entire 2:00-3:00 partition, or are you intending to scan the records that have currently dropped in the 3:00 partition so far? I think it varies from query to query...

Reply:
Not an easy problem to solve